terra_notebook_utils/blobstore/local.py

```diff
@@ -0,0 +1,117 @@
+import os
+import shutil
+from math import ceil
+from functools import wraps
+from typing import Dict, Generator, Optional
+
+from getm import default_chunk_size
+
+from terra_notebook_utils import blobstore
+
+
+def catch_blob_not_found(func):
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        try:
+            return func(self, *args, **kwargs)
+        except FileNotFoundError as ex:
+            raise blobstore.BlobNotFoundError(f"Could not find {self.url}") from ex
+    return wrapper
+
+class LocalBlobStore(blobstore.BlobStore):
+    def __init__(self, basepath: str):
+        self.bucket_name = basepath
+
+    def list(self, prefix: str="") -> Generator["LocalBlob", None, None]:
+        root = os.path.join(self.bucket_name, prefix)
+        if root.endswith(os.path.sep):
+            root = root[:-len(os.path.sep)]
+        for (dirpath, dirnames, filenames) in os.walk(root):
+            for filename in filenames:
+                relpath = os.path.relpath(os.path.join(dirpath, filename), self.bucket_name)
+                yield LocalBlob(self.bucket_name, relpath)
+
+    def blob(self, key: str) -> "LocalBlob":
+        return LocalBlob(self.bucket_name, key)
+
+class LocalBlob(blobstore.Blob):
+    def __init__(self, basepath: str, relpath: str):
+        assert basepath == os.path.abspath(basepath)
+        self.bucket_name = basepath
+        self.key = relpath
+        self._path = os.path.join(basepath, relpath)
+
+    # The next two methods customize pickling behavior
+    def __getstate__(self):
+        return dict(bucket_name=self.bucket_name, key=self.key, _path=self._path)
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+
+    @property
+    def url(self) -> str:  # type: ignore # TODO https://github.com/python/mypy/issues/4125
+        return self._path
+
+    @catch_blob_not_found
+    def get(self) -> bytes:
+        with open(self._path, "rb") as fh:
+            return fh.read()
+
+    def put(self, data: bytes):
+        with open(self._path, "wb") as fh:
+            fh.write(data)
+
+    @catch_blob_not_found
+    def delete(self):
+        os.remove(self._path)
+
+    @catch_blob_not_found
+    def copy_from(self, src_blob: "LocalBlob"):
+        """
+        Intra-cloud copy
+        """
+        assert isinstance(src_blob, type(self))
+        if self.url != src_blob.url:
+            shutil.copyfile(src_blob._path, self._path)
```

Comment:
Is there a reason to avoid the `SameFileError`?  Just curious if this tries to copy an object to itself very often.

Reply:
Why bother with SameFileError? Seems easier to just do nothing in this case.